xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-io/src/main/java/org/xwiki/annotation/io/internal/migration/hibernate/R160000000XWIKI20699DataMigration.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.annotation.io.internal.migration.hibernate;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.internal.migration.AbstractDocumentsMigration;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.query.Query;
+import org.xwiki.query.QueryException;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.store.migration.DataMigrationException;
+import com.xpn.xwiki.store.migration.XWikiDBVersion;
+
+/**
+ * Select and queue the document that can require an annotation target cleanup. Self-referencing annotation targets are
+ * not necessary and can get outdated in case of page move/copy. Therefore, it is better to replace them with the empty
+ * string, implicitly targeting the document holding the annotation XObject.
+ *
+ * @version $Id$
+ * @since 15.5.5
+ * @since 15.10.1
+ * @since 16.0RC1
+ */
+@Component
+@Named("R160000000XWIKI20699")
+@Singleton
+public class R160000000XWIKI20699DataMigration extends AbstractDocumentsMigration
+{
+    @Override
+    public String getDescription()
+    {
+        return "Queue documents containing annotations with a non-empty target for cleanup";
+    }
+
+    @Override
+    public XWikiDBVersion getVersion()
+    {
+        return new XWikiDBVersion(160000000);
```

Comment:
We need to be careful, that one already exists in master: https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-default/src/main/java/org/xwiki/notifications/filters/migration/R160000000XWIKI17243DataMigration.java#L122